### ISSUE TITLE
HttpClient.PostAsync throws HttpRequestException when Keep-Alive'd connection is remotely closed due to idle timeout

### ISSUE THREAD

**Issue by dmitryvk** [mono/mono#17710]
I've noticed that non-GET requests in HttpClient will fail with HttpRequestException if remote server closes pooled connection.

## Steps to Reproduce

1. Prepare an HTTP server that uses Keep-Alive but has small idle timeout (1 second).

    * For example, use `run_nginx.sh` script from
https://gist.github.com/dmitryvk/2bf190d41267b242ed8ca31f63ea40e0#file-run_nginx-sh. This script will run minimal nginx server which has Keep-Alive idle timeout set to 1 second (note `keepalive_timeout 1;` in nginx_template.conf).

2. Compile and run the following program: https://gist.github.com/dmitryvk/2bf190d41267b242ed8ca31f63ea40e0#file-program-cs

### Current Behavior

Programs outputs:

```
<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.12.1</center>
</body>
</html>

System.AggregateException: One or more errors occurred. (An error occurred while sending the request.) ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00cf5] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x012e5] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync (System.Net.Http.HttpConnection connection, System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0012b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithRetryAsync (System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0014b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.RedirectHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000ba] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpClient.FinishSendAsyncBuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x0017e] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at System.Threading.Tasks.Task`1[TResult].GetResultCore (System.Boolean waitCompletionNotification) [0x0002b] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at System.Threading.Tasks.Task`1[TResult].get_Result () [0x0000f] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at HttpClientTest.Program.Main (System.String[] args) [0x0005c] in <d73b3300d48b402092ac95202adc3c1c>:0 
---> (Inner Exception #0) System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00cf5] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x012e5] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync (System.Net.Http.HttpConnection connection, System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0012b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithRetryAsync (System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0014b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.RedirectHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000ba] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpClient.FinishSendAsyncBuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x0017e] in <f060cf3977d54c358d58805336bde9d1>:0 <---

```

### Expected Behavior

Program should output (as it does when run in .NET Core):

```
<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.12.1</center>
</body>
</html>

<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.12.1</center>
</body>
</html>

```

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 6.4.0.198 (tarball Mon Sep 23 20:53:41 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug 
        Interpreter:   yes
        LLVM:          yes(610)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

## Stacktrace

```
System.AggregateException: One or more errors occurred. (An error occurred while sending the request.) ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00cf5] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x012e5] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync (System.Net.Http.HttpConnection connection, System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0012b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithRetryAsync (System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0014b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.RedirectHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000ba] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpClient.FinishSendAsyncBuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x0017e] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at System.Threading.Tasks.Task`1[TResult].GetResultCore (System.Boolean waitCompletionNotification) [0x0002b] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at System.Threading.Tasks.Task`1[TResult].get_Result () [0x0000f] in <6ea21d94575948579c2890f43ad5bec5>:0 
  at HttpClientTest.Program.Main (System.String[] args) [0x0005c] in <d73b3300d48b402092ac95202adc3c1c>:0 
---> (Inner Exception #0) System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00cf5] in <f060cf3977d54c358d58805336bde9d1>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x012e5] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync (System.Net.Http.HttpConnection connection, System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0012b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpConnectionPool.SendWithRetryAsync (System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0014b] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.RedirectHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000ba] in <f060cf3977d54c358d58805336bde9d1>:0 
  at System.Net.Http.HttpClient.FinishSendAsyncBuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x0017e] in <f060cf3977d54c358d58805336bde9d1>:0 <---
```

**Comment by dmitryvk** [comment]
Here's result of packet capture:
[mono-http-client-keepalive-idle-timeout.zip](https://github.com/mono/mono/files/3808948/mono-http-client-keepalive-idle-timeout.zip)
![image](https://user-images.githubusercontent.com/120884/68207921-1960a780-ffe1-11e9-9609-96626be8dec0.png)

Please not that the connection was closed by nginx at around 1.0 second (packet 8), but application wrote to this conneciton 0.5 seconds later (packet 10).

**Comment by dmitryvk** [comment]
I've traced the problem down to System.Net.Http.HttpConnection.EnsureReadAheadAndPollRead (https://github.com/dotnet/corefx/blob/97d9fb9d6c7e1d50f527e290ef8e18767dbc6bbf/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L170). This methods is supposed to return `true` if connection is either closed or has some data.

On Mono, EnsureReadAheadAndPollRead returns `false` for remotely closed connection while on .NET Core this method returns `true`. This happens due to erroneous assumption that NetworkStream.ReadAsync returns *synchronously completed* Task (i.e., that ReadAsync of zero bytes is equivalent to `select` or `poll` syscall). see https://gist.github.com/dmitryvk/2bf190d41267b242ed8ca31f63ea40e0#file-program_networkstreamsynccompleted-cs: on Mono, this prints
```
Connected
Accepted and closed
IsSyncCompleted initial: False
IsSyncCompleted after sleep: True
ReadAsync result: 0

```
while on .NET Core this prints:
```
Connected
Accepted and closed
IsSyncCompleted initial: True
IsSyncCompleted after sleep: True
ReadAsync result: 0

```

**Comment by baulig** [comment]
I can reproduce this every single time - see my comments on the corefx issue.

**Comment by baulig** [comment]
Should have been fixed in #18275
